raiden/tests/integration/long_running/test_settlement.py

```diff
@@ -481,6 +483,114 @@ def run_test_channel_withdraw(
     assert bob_initial_balance + total_withdraw == bob_balance_after_withdraw
 
 
+@pytest.mark.parametrize("number_of_nodes", [2])
+def test_channel_withdraw_expired(
+    raiden_network, number_of_nodes, token_addresses, deposit, network_wait, retry_timeout
+):
+    raise_on_failure(
+        raiden_network,
+        run_test_channel_withdraw_expired,
+        raiden_network=raiden_network,
+        token_addresses=token_addresses,
+        deposit=deposit,
+        network_wait=network_wait,
+        number_of_nodes=number_of_nodes,
+        retry_timeout=retry_timeout,
+    )
+
+
+def run_test_channel_withdraw_expired(
+    raiden_network, token_addresses, deposit, network_wait, number_of_nodes, retry_timeout
+):
+    """ Tests withdraw expiration. """
+    alice_app, bob_app = raiden_network
+    token_address = token_addresses[0]
+    token_network_address = views.get_token_network_address_by_token_address(
+        views.state_from_app(alice_app), alice_app.raiden.default_registry.address, token_address
+    )
+
+    message_handler = WaitForMessage()
+    bob_app.raiden.message_handler = message_handler
+
+    hold_event_handler = HoldRaidenEventHandler(RaidenEventHandler())
+    # Prevent withdraw confirmation from being sent
+    send_withdraw_confirmation_event = hold_event_handler.hold(SendWithdrawConfirmation, {})
+
+    alice_app.raiden.raiden_event_handler = hold_event_handler
+    bob_app.raiden.raiden_event_handler = hold_event_handler
+
+    alice_to_bob_amount = 10
+    identifier = 1
+    target = bob_app.raiden.address
+    secret = sha3(target)
+
+    payment_status = alice_app.raiden.start_mediated_transfer_with_secret(
+        token_network_address=token_network_address,
+        amount=alice_to_bob_amount,
+        fee=0,
+        target=target,
+        identifier=identifier,
+        secret=secret,
+    )
+
+    wait_for_unlock = bob_app.raiden.message_handler.wait_for_message(
+        Unlock, {"payment_identifier": identifier}
+    )
+    timeout = network_wait * number_of_nodes
+    with Timeout(seconds=timeout):
+        wait_for_unlock.get()
```

Comment:
For some reason, waiting for `payment_done` is not enough. I'll have to check why this is the case.

Reply:
Ahh... alice sends the payment and when the secret is revealed, alice sets the `payment_done` while bob hasn't received the `Unlock` message yet. This could lead to flakiness so i think it's better to keep bob waiting for the unlock to make sure the withdraw happens on the correct balance **after** the transfer is complete.